README.md
# MyBBBridge

MyBBBridge is an extension aimed at making MyBB template and theme dev more convenant.

**Warning: this software is pre-alpha. Make backups and use it at your own risk.**

## Features

* Download themes (`.css`) and templates (`.html`) files from database to disk.

* Save themes and templates to database when saved on disk.

* Automatically refresh templates cache files on save.

## How to use

### Configuration

To operate, MyBBBridge *needs* a `.vscode/mbbb.json` file at your workspace root.

You can create one using the command `MyBBBridge: Create config file`.

The default config file looks like this:

```json
{
    "database": {
        "host": "localhost",
        "port": 3306,
        "database": "mybb",
        "prefix": "mybb_",
        "user": "root",
        "password": ""
    },
    "mybbVersion": 1860,
    "mybbUrl": "http://localhost",
    "autoUpload": true
}
```

* `database`: This one should be quite self explanatory!

* `mybbVersion`: MyBB version to be used in newly created theme files. Existing files
  will keep their version metadata.

* `mybbUrl`: URL of your MyBB board. Set to `null` or `''` to disable cache refresh
  requests.

* `autoUpload`: If true, MyBBBridge will try to save theme and stylesheets to database
  each time a corresponding file is saved in VSCode.
  *Overrides existing database entries without confirmation!*

### Cache refresh

To be able to ask MyBB for template cache refresh, MyBBBridge requires you to upload
the tiny `cachecss.php` php file of this repository to your web server, at the root of
your MyBB directory. If you don't plan to use cache refresh, you can skip this step and
set `mybbUrl` to `null` in your `mbbb.json` config file.

### Commands

* `MyBBBridge: Create config file`: Create a new config file, allowing you to start
  using MyBBBridge.

* `MyBBBridge: Load MyBB template set from database`: Download and save all templates
  files of a given template set to the `./template_sets/<template_set_name>/` folder.
  *Overrides existing files without confirmation!*

* `MyBBBridge: Load MyBB style from database`: Download and save all stylesheet files
  of a given style to the `./styles/<style_name>/` folder.
  *Overrides existing files without confirmation!*

## Release Notes

### 0.0.1-alpha

Alpha release providing basic download features.

### 0.0.2-alpha

Alpha release with save features.

### 0.0.3-alpha

Alpha release allowing to refresh MyBB stylesheet cache files.


CHANGELOG.md
# Change Log

All notable changes to the "mybbbridge" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

src/utils.ts
import * as vscode from 'vscode';
import * as mysql from 'mysql';
import { promises as fs, PathLike } from 'fs';
import path = require('path');


export function timestamp(): number {
    return Math.floor(Date.now() / 1000);
}


export function getWorkspacePath(): string {
    if (vscode.workspace.workspaceFolders) {
        return vscode.workspace.workspaceFolders[0].uri.fsPath;
    } else {
        vscode.window.showErrorMessage("No workspace opened!");
    }
    return '';
}


export async function makePath(path: PathLike): Promise<void> {
    try {
        await fs.mkdir(path, { 'recursive': true });
    } catch (err) {
        if (err.code !== 'EEXIST') {
            throw err;
        }
    }
}


export function urlJoin(urlParts: string[]): string {
    return urlParts
        .map(part => {
            const part2 = part.endsWith('/') ? part.substring(0, part.length - 1) : part;
            return part2.startsWith('/') ? part2.substr(1) : part2;
        })
        .join('/');
}


export async function getConfig(): Promise<any> {
    const configFilePath = path.join(getWorkspacePath(), '.vscode', 'mbbb.json');
    let configFile: Buffer;
    try {
        configFile = await fs.readFile(configFilePath);
    } catch (err) {
        if (err.code === 'ENOENT') {
            vscode.window.showErrorMessage('Config file not found! Try using "Create config file" command.');
            return;
        }
        throw err;
    }
    return JSON.parse(configFile.toString());
}


export function getConnexion(dbConfig: any): mysql.Connection {
    return mysql.createConnection({
        host: dbConfig.host,
        port: dbConfig.port,
        database: dbConfig.database,
        user: dbConfig.user,
        password: dbConfig.password
    });
}

src/events.ts
import * as vscode from 'vscode';
import path = require('path');

import { MyBBTemplateSet, MyBBStyle } from "./MyBBThemes";
import { getWorkspacePath, getConfig, getConnexion } from './utils';

export async function onSaveEvent(document: vscode.TextDocument) {
    const config = await getConfig();

    if (config.autoUpload) {
        const docPath = document.uri.fsPath;
        const parent1Path = path.dirname(docPath);
        const parent2Path = path.dirname(parent1Path);
        const parent3Path = path.dirname(parent2Path);

        if (parent3Path === getWorkspacePath()) {
            const ext = path.extname(docPath);
            const parent1Dir = path.basename(parent1Path);
            const parent2Dir = path.basename(parent2Path);

            const con = getConnexion(config.database);

            if (parent2Dir === 'template_sets' && ext === '.html') {
                const templateSet = new MyBBTemplateSet(parent1Dir, con, config.database.prefix);
                const fileName = path.basename(docPath, ext); // With extension
                templateSet.saveElement(fileName, document.getText(), config.mybbVersion);

            } else if (parent2Dir === 'styles' && ext === '.css') {
                const style = new MyBBStyle(parent1Dir, con, config.database.prefix);
                const fileName = path.basename(docPath); // Without extension
                style.saveElement(fileName, document.getText(), config.mybbVersion);
            }
        }
    }
    
}

src/MyBBThemes.ts
import * as vscode from 'vscode';
import * as mysql from 'mysql';
import * as request from 'request-promise-native';

import { timestamp, urlJoin, getConfig } from './utils';


abstract class MyBBSet {
    name: string;
    con: mysql.Connection;
    prefix: string;

    public constructor(name:string, con: mysql.Connection, prefix: string='mybb_') {
        this.name = name;
        this.con = con;
        this.prefix = prefix;
    }

    public getTable(name:string): string {
        return this.prefix + name;
    }

    public query(req: string, params: any[], callback: any=()=>{}): any {
        return new Promise((resolve, reject) => {
            this.con.query(req, params, (err: any, result: any) => {
                if (err) {
                    vscode.window.showErrorMessage(err.message);
                    return reject(err);
                }
                callback(err, result);
                return resolve(result);
            });
        });
    }
}


export class MyBBTemplateSet extends MyBBSet {
    sid: number|undefined;

    private async getSid(): Promise<number|undefined> {
        const res = await this.query(
            'SELECT sid FROM ?? WHERE title=?',
            [this.getTable('templatesets'), this.name],
            (err: any, result: any) => {
                if (!result.length) {
                    vscode.window.showErrorMessage(`Can't find template ${this.name}!`);
                }
            }
        );
        this.sid = res[0].sid;
        return this.sid;
    }

    public async getElements(): Promise<any> {
        await this.getSid();
        const templates = await this.query(
            'SELECT title, template FROM ?? WHERE sid=? ORDER BY sid DESC, title ASC',
            [this.getTable('templates'), this.sid],
            (err: any, result: any) => {
                if (!result.length) {
                    vscode.window.showErrorMessage(`No templates files found for template ${this.name}!`);
                } 
            }
        );
        return templates;
    }

    public async saveElement(name: string, content: string, version: number) {
        await this.getSid();

        const res = await this.query(
            'SELECT tid FROM ?? WHERE title=? AND sid=?',
            [this.getTable('templates'), name, this.sid],
        );

        if (!res.length) { // Insert
            this.query(
                'INSERT INTO ?? SET title=?, template=?, sid=?, version=?',
                [this.getTable('templates'), name, content, this.sid, version],
                (err: any, result: any) => {
                    if (!err) {
                        vscode.window.showInformationMessage(`Uploaded new template ${name} in DB.`);
                    }
                }
            );
        } else { // Update
            this.query(
                'UPDATE ?? SET template=? WHERE title=? AND sid=?',
                [this.getTable('templates'), content, name, this.sid],
                (err: any, result: any) => {
                    if (!err) {
                        vscode.window.showInformationMessage(`Updated template ${name} in DB.`);
                    }
                }
            );
        }
    }
}


export class MyBBStyle extends MyBBSet {
    tid: number|undefined;

    private async getTid() {
        const res = await this.query(
            'SELECT tid FROM ?? WHERE name=?',
            [this.getTable('themes'), this.name],
            (err: any, result: any) => {
                if (!result.length) {
                    vscode.window.showErrorMessage(`Can't find theme ${this.name}!`);
                }
            }
        );
        this.tid = res[0].tid;
        return this.tid;
    }

    private async getSid(name: string) {
        await this.getTid();
        const res = await this.query(
            'SELECT sid FROM ?? WHERE name=? AND tid=?',
            [this.getTable('themestylesheets'), name, this.tid]
        );
        return res[0].sid;
    }

    public async getElements() {
        await this.getTid();
        const stylesheets = await this.query(
            'SELECT name, stylesheet FROM ?? WHERE tid=? ORDER BY tid DESC, name ASC',
            [this.getTable('themestylesheets'), this.tid],
            (err: any, result: any) => {
                if (!result.length) {
                    vscode.window.showErrorMessage(`No stylesheets files found for theme ${this.name}!`);
                } 
            }
        );
        return stylesheets;
    }

    public async saveElement(name: string, content: string, version: number) {
        await this.getTid();

        const res = await this.query(
            'SELECT sid FROM ?? WHERE name=? AND tid=?',
            [this.getTable('themestylesheets'), name, this.tid],
        );

        if (!res.length) { // Insert
            this.query(
                'INSERT INTO ?? SET name=?, stylesheet=?, tid=?, lastmodified=?',
                [this.getTable('themestylesheets'), name, content, this.tid, timestamp()],
                (err: any, result: any) => {
                    if (!err) {
                        vscode.window.showInformationMessage(`Uploaded new stylesheet ${name} in DB.`);
                    }
                }
            );
        } else { // Update
            this.query(
                'UPDATE ?? SET stylesheet=?, lastmodified=? WHERE name=? AND tid=?',
                [this.getTable('themestylesheets'), content, timestamp(), name, this.tid],
                (err: any, result: any) => {
                    if (!err) {
                        vscode.window.showInformationMessage(`Updated stylesheet ${name} in DB.`);
                        this.requestCacheRefresh(name);
                    }
                }
            );
        }
    }

    public async requestCacheRefresh(name: string): Promise<void> {
        const config = await getConfig();

        if (config.mybbUrl) {
            const scriptUrl = urlJoin([config.mybbUrl, 'cachecss.php']);

            request
                .get({
                    uri: scriptUrl,
                    qs: {
                        tid: this.tid,
                        name: name
                    }
                })
                .catch(err => {
                    vscode.window.showErrorMessage("Failed to request stylesheet cache refresh!");
                    throw err;
                });
        }
    }
}

src/loadCommands.ts
import * as vscode from 'vscode';
import { promises as fs } from 'fs';
import path = require('path');

import { MyBBTemplateSet, MyBBStyle } from "./MyBBThemes";
import { getWorkspacePath, makePath, getConfig, getConnexion } from './utils';


export async function loadTemplateSetCommand() {
    const config = await getConfig();
    const con = getConnexion(config.database);

    const templateSetName = await vscode.window.showInputBox({ placeHolder: 'Template set name (often ending with "Templates")' });
    if (templateSetName === undefined) {
        return;
    }
    const templateSet = new MyBBTemplateSet(templateSetName, con, config.database.prefix);

    const templateSetPath = path.join(getWorkspacePath(), 'template_sets', templateSet.name);
    await makePath(templateSetPath);

    const templates = await templateSet.getElements();
    templates.forEach(async (template: any) => {
        let templatePath = path.join(templateSetPath, template.title + '.html');
        await fs.writeFile(templatePath, template.template);
    });
    vscode.window.showInformationMessage(`${templates.length} templates were loaded.`);
}


export async function loadStyleCommand() {
    const config = await getConfig();
    const con = getConnexion(config.database);

    const styleName = await vscode.window.showInputBox({ placeHolder: 'Style name' });
    if (styleName === undefined) {
        return;
    }
    const style = new MyBBStyle(styleName, con, config.database.prefix);

    const stylePath = path.join(getWorkspacePath(), 'styles', style.name);
    await makePath(stylePath);

    const stylesheets = await style.getElements();
    stylesheets.forEach(async (stylesheet: any) => {
        let stylesheetPath = path.join(stylePath, stylesheet.name);
        await fs.writeFile(stylesheetPath, stylesheet.stylesheet);
    });
    vscode.window.showInformationMessage(`${stylesheets.length} stylesheets were loaded.`);
}

src/utilCommands.ts
import * as vscode from 'vscode';
import { promises as fs, PathLike } from 'fs';
import path = require('path');

import { getWorkspacePath } from './utils';


export async function createConfigCommand() {
    // Get current workspace path
    const workspacePath = getWorkspacePath();
    
    // Get or create .vscode dir
    let configFilePath = path.join(workspacePath, '.vscode');
    try {
        await fs.mkdir(configFilePath);
    } catch (err) {
        if (err.code !== 'EEXIST') {
            throw err;
        }
    }

    // Create config file if not already existing
    configFilePath = path.join(configFilePath, 'mbbb.json');
    try {
        await fs.access(configFilePath);
    } catch (err) {
        if (err.code === 'ENOENT') {
            const defaultConf = JSON.stringify({
                "database": {
                    "host": "localhost",
                    "port": 3306,
                    "database": "mybb",
                    "prefix": "mybb_",
                    "user": "root",
                    "password": ""
                },
                "mybbVersion": 1860,
                "mybbUrl": "http://localhost",
                "autoUpload": true
            }, null, 4);

            await fs.writeFile(configFilePath, defaultConf);
            vscode.window.showInformationMessage(`Config file ${configFilePath} created sucessfully.`);
            return;
        }
        throw err;
    }
    vscode.window.showErrorMessage(`Config file ${configFilePath} already exists!`);
}